Allow user to add metadata to config section "user"

PinkShnack opened the pull request:
Attempt at a solution to #125. Will close #125 

This may not be the sleekest method, but it works and just required me to set some exclusions for the "user" section of the config file.

### To do
- [x] be able to save and reload new "user" section of config file
- [x] add tests for user config section
   - [x] general tests (need to verify what should be allowed)
   - [x] save and reload test
   - [x] test with plugin and temporary features
   - [x] test docstrings and correct order
 - [x] add documentation section (specifically mention on plugins page also)
  
Regarding the general tests, we may want to limit the user to only certain data types.
~~For example, one can currently set `ds.config["user"]["something"] = [2, "a word", True, "another word", 4.4]`. Do we want to restrict this information to being `str`, `bool`, `int`, float`?~~

EDIT:

The keys for the "user" section will be resticted to strings (`str`). The values can be other data structures (string, numbers, list, numpy array).

felix-r commented:
I skimmed through the `TrackedDataSet` attributes and one possible metadata would come in list format as a list of tuples. I could create a workaround for this but just sayin'.

paulmueller commented on tests/test_plugin_features.py:
```diff
@@ -476,6 +476,34 @@ def test_pf_with_no_feature_label():
+    """Use a plugin feature with the user defined config section"""
+    # setup a plugin method that uses user config section
+    def compute_with_user_section(rtdc_ds):
+        if rtdc_ds.config["user"]["channel"]:
```
Here you should also check whether it `isinstance` of `bool`, because ``bool("true") == True``.

paulmueller commented on tests/test_plugin_features.py:
```diff
@@ -476,6 +476,34 @@ def test_pf_with_no_feature_label():
+
+    info = {"method": compute_with_user_section,
+            "feature names": ["area_of_region"]}
+    _ = PlugInFeature("area_of_region", info)
```
General remark: If you IDE complains if you omit `_ = `, then I would say it is ok to leave it there. Personally I think it makes the code less readable.

paulmueller reviewed (changes requested):
:heart:

In addition to my comments, I have a few other points:

1. Please make sure that dataset checks always ignore the "user" config (see test_rtdc_check_dataset.py) - I don't know whether this is handled properly currently.
2. There should be a separate section in the docs on user-defined configuration keys and the docs on plugin features should explicitly mention that section (I believe plugin features are one key function where the "user" config is needed).

PinkShnack commented on tests/test_rtdc_config.py:
```diff
@@ -81,6 +83,115 @@ def test_config_update():
+def test_user_section_key_types_unhashable():
+    """Show the user section keys that are not allowed (unhashable)"""
+    ds = new_dataset(retrieve_data("rtdc_data_hdf5_rtfdc.zip"))
+    with pytest.raises(TypeError):
```
Ah good catch

PinkShnack commented on dclab/definitions.py:
```diff
@@ -165,6 +165,10 @@
     ],
+    # All user-defined metadata keywords
+    "user": [
+        ["example", str, "some example metadata"],
```
Indeed this (and the below comment) would be great. I also don't like having it there. However, it does I think make this a bit needlessly complicated. See ecd4b95129db0564938602f67ab1f8b502bd15e3 to see all the things I had to add...

I will try to just remove the example line and see if it works...

paulmueller commented on dclab/rtdc_dataset/fmt_hdf5.py:
```diff
@@ -198,6 +198,9 @@ def __init__(self, h5path, *args, **kwargs):
 
         # Parse configuration
         self.config = RTDC_HDF5.parse_config(h5path)
+        # Add "user" section if it doesn't exist
```
Adding the user section should not be necessary here. This must be handled globally in config.py `Configuration.__getitem__` (Otherwise user sections won't work with e.g. DCOR or dictionary data).

paulmueller reviewed (changes requested):
- Please add tests for all RT-DC data formats to see whether setting user-defined configuration and exporting to HDF5 works with:
  - `fmt_dcor.RTDC_DCOR`
  - `fmt_dict.RTDC_Dict`
  - `fmt_tdms.RTDC_TDMS`

- Does exporting an empty "user" configuration to HDF5 work?

PinkShnack commented on dclab/rtdc_dataset/write_hdf5.py:
```diff
@@ -240,15 +240,15 @@ def write(path_or_h5file, data=None, meta=None, logs=None, mode="reset",
     # Check meta data
     for sec in meta:
-        if sec not in dfn.CFG_METADATA:
+        if sec == "user":
```
[Us watching the users struggle](https://external-content.duckduckgo.com/iu/?u=https%3A%2F%2Fmedia.giphy.com%2Fmedia%2F3oz8xDW2VU5qHu6EJG%2Fgiphy.gif&f=1&nofb=1)

PinkShnack commented on tests/test_rtdc_config.py:
```diff
@@ -97,17 +97,18 @@ def test_user_section_basic():
 def test_user_section_different_key_types():
     """Check that the user config section keys take different data types"""
```
I agree, keys should be strings. This is what you see in HDFView with the following code:
```python
ds.config["user"][True] = True
ds.config["user"][23.5] = 23.5
ds.config["user"][12] = 12
ds.config["user"]["a string"] = "a string"
ds.config["user"][(5, 12.3, False, "a word")] = "a word"
```

![config_non-string-inputs](https://user-images.githubusercontent.com/43960046/120314429-87637280-c2db-11eb-8462-d6237d54119d.PNG)

PinkShnack commented on tests/test_rtdc_config.py:
```diff
@@ -81,6 +84,267 @@ def test_config_update():
+
+
+# def test_user_section_set_save_reload_fmt_dict():
+#     """Check that 'user' section metadata works for RTDC_Dict"""
```
I can't used the `with` context manager here, so the file is output each time and `override` is needed. This is okay but not perfect.

paulmueller replied:
You should create "exported.rtdc" in a temporary directory.

PinkShnack commented on tests/test_rtdc_config.py:
```diff
@@ -81,6 +84,267 @@ def test_config_update():
+#
+#
+# def test_user_section_set_save_reload_fmt_dcor():
+#     """Check that 'user' section metadata works for RTDC_Dcor"""
```
This test hangs so need to do more checks on it.

paulmueller replied:
Probably hangs because it downloads and exports *all* features. Just use e.g. `features=["area_um", "deform"]`.

PinkShnack commented:
Hey @paulmueller, I have added some documentation and tests. Will you please review them? Cheers!

paulmueller commented:
Please make sure that all the tests run through and that the docs build successfully (see my other comment).

PinkShnack commented:
I get some nice warnings from sphinx. Sphinx usually has lots of warnings that one can ignore. Are these ones normal:

```
fakefile.rst:16: WARNING: image file not readable: ../examplesoverview_plot.jpg
fakefile.rst:19: WARNING: image file not readable: ../examplesemodulus_dcor.jpg
fakefile.rst:24: WARNING: image file not readable: ../exampleslme4_lmer.jpg
fakefile.rst:33: WARNING: image file not readable: ../examplesml_tensorflow.jpg
fakefile.rst:37: WARNING: image file not readable: ../examplesml_builtin.jpg
fakefile.rst:10: WARNING: image file not readable: ../examplesisoelastics.jpg
fakefile.rst:13: WARNING: image file not readable: ../examplesplugin_usage.jpg
C:\Users\eoconne\Documents\Python\repos\dclab\docs\sec_examples.rst:14: WARNING: download file not readable: C:\Users\eoconne\Documents\Python\repos\dclab\examplesoverview_plot.py
C:\Users\eoconne\Documents\Python\repos\dclab\docs\sec_examples.rst:17: WARNING: download file not readable: C:\Users\eoconne\Documents\Python\repos\dclab\examplesemodulus_dcor.py
C:\Users\eoconne\Documents\Python\repos\dclab\docs\sec_examples.rst:22: WARNING: download file not readable: C:\Users\eoconne\Documents\Python\repos\dclab\exampleslme4_lmer.py
C:\Users\eoconne\Documents\Python\repos\dclab\docs\sec_examples.rst:15: WARNING: download file not readable: C:\Users\eoconne\Documents\Python\repos\dclab\exampleslme4_glmer_diff.py
fakefile.rst:34: WARNING: download file not readable: C:\Users\eoconne\Documents\Python\repos\dclab\examplesml_tensorflow.py
fakefile.rst:38: WARNING: download file not readable: C:\Users\eoconne\Documents\Python\repos\dclab\examplesml_builtin.py
C:\Users\eoconne\Documents\Python\repos\dclab\docs\sec_examples.rst:8: WARNING: download file not readable: C:\Users\eoconne\Documents\Python\repos\dclab\examplesisoelastics.py
C:\Users\eoconne\Documents\Python\repos\dclab\docs\sec_examples.rst:11: WARNING: download file not readable: C:\Users\eoconne\Documents\Python\repos\dclab\examplesplugin_usage.py

```

PinkShnack commented:
I have some nice code for you:
``` python
if sphinx_warnings == "okay" and appveyor == "passes" and Paul.review == "a happy chappy":
    merge.please() 
```
You are now a python class, get used to your new life :')

paulmueller commented:
The warnings you are getting are WIndows-related (path-joining). I wrote the fancy_include sphinx extension on Linux and never tested it on Windows.

paulmueller commented on tests/test_rtdc_config.py:
```diff
@@ -81,6 +85,284 @@ def test_config_update():
+    with h5py.File(expath, "r") as h5:
+        assert h5.attrs["user:some metadata"] == 42
+    # fails due to RTDC_HDF5 trying to add self.title
+    with pytest.raises(KeyError):
```
Please change this test so that this works (no KeyError). i.e. add `ds.config["experiment"]["sample"]="test"` and `ds.config["experiment"]["run index"]=1` before exporting.

PinkShnack replied:
Ah, of course...

paulmueller commented on tests/test_rtdc_config.py:
```diff
@@ -81,6 +85,284 @@ def test_config_update():
+            assert ds2.config["user"] == metadata
+
+
+def test_user_section_set_save_reload_fmt_dcor():
```
Please add decorator `@pytest.mark.skipif(not DCOR_AVAILABLE, reason="DCOR not reachable!")` (see `test_rtdc_fmt_dcor.py`).

PinkShnack replied:
I went through this and then somehow came to the conclusion that I wouldn't use it. :')

paulmueller commented on tests/test_rtdc_config.py:
```diff
@@ -81,6 +85,284 @@ def test_config_update():
+        assert all(h5.attrs["user:outlet"] == outlet)
+    # now check again with dclab
+    with new_dataset(expath) as ds2:
+        for k1, k2 in zip(ds2.config["user"], metadata):
```
if `k1 == k2`, then this for loop could be simplified without `zip`.

paulmueller reviewed (changes requested):
Just a few minor things that need sorting-out. Nicely done!

paulmueller commented on docs/sec_av_notation.rst:
```diff
@@ -178,8 +178,75 @@ The following are supported:
+
+        # we can clear the "user" section like so:
+        In [9]: ds.config["user"] = {}
+
```
Here, please encourage users to use str, bool, float, and int, because containers might not render nicely in Shape-Out or DCOR.

paulmueller commented on dclab/rtdc_dataset/check.py:
```diff
@@ -97,7 +97,8 @@
         "bg empty",
-    ]
+    ],
+    "user": [],
```
I doubt that this is enough.

PinkShnack replied:
I reckon this is something I need an overview on (how these checks work)

PinkShnack replied:
Just seeing the below comment, will look into that now.

paulmueller commented:
Wait, I found two more things:

1. Please add a dataset check test with user-defined features. You can start with this test: https://github.com/ZELLMECHANIK-DRESDEN/dclab/blob/master/tests/test_rtdc_check_dataset.py#L47 and just add user-defined features before `check_dataset`.
2. Some of the new test functions emit (expected) warnings that are partially caused e.g. by the tdms file format (InitialFrameMissingWarning, LimitingExportSizeWarning). These should be explicitly ignored via a decorator added to the test function.

PinkShnack commented on docs/sec_av_notation.rst:
```diff
@@ -178,8 +178,75 @@ The following are supported:
+        In [8]: print(ds.config["user"])
+
+        # we can clear the "user" section like so:
+        In [9]: ds.config["user"] = {}
```
Indeed it does (woohoo new tests!) I will replace `ds.config["user"] = {}` with `ds.config["user"].clear()`

PinkShnack commented on tests/test_rtdc_config.py:
```diff
@@ -81,6 +85,284 @@ def test_config_update():
+        assert ds2.config["user"] == {}
+
+
+def test_user_section_set_save_reload_empty_key():
```
I actually didn't this time... :D

PinkShnack commented on tests/test_rtdc_config.py:
```diff
@@ -81,6 +85,284 @@ def test_config_update():
+
+
+@pytest.mark.parametrize("user_config", [{"": ""}, {" ": ""}])
+def test_user_section_set_save_reload_fails(user_config):
```
Welcome to the word of empty strings!
Well `config.py` does not allow the value to be an empty string. That shows why `{"": ""}` and `{" ": ""}` are not allowed. So these cases are already covered. 

We could indeed add a check for `{"": " "}`. It shouldn't be allowed...

But what about `{" ": " "}`? Will we go to the bother of checking that there are more than one character that isn't a single space?

paulmueller replied:
I think it would be safe to at least to a `.strip()` for every key (not for the value) and make sure the result is not empty. This check is a convenience check for users to prevent bad things from happening.

PinkShnack commented on docs/sec_av_notation.rst:
```diff
@@ -178,8 +178,73 @@ The following are supported:
+
+        In [2]: ds = dclab.new_dataset("data/example.rtdc")
+
+        In [3]: my_metadata = {"inlet": True, n_channels: 4}
```
I think you might have been commenting on an older commit. This had been since fixed!

PinkShnack commented:
> Wait, I found two more things:
> 
> 1. Please add a dataset check test with user-defined features. You can start with this test: https://github.com/ZELLMECHANIK-DRESDEN/dclab/blob/master/tests/test_rtdc_check_dataset.py#L47 and just add user-defined features before `check_dataset`.
> 2. Some of the new test functions emit (expected) warnings that are partially caused e.g. by the tdms file format (InitialFrameMissingWarning, LimitingExportSizeWarning). These should be explicitly ignored via a decorator added to the test function.

1. Implemented this, but is just ignoring "user" in `IntegrityChecker.check_metadata_bad` sufficient? See the changes [here](https://github.com/PinkShnack/dclab/blob/7d3021e9e59b769d056ffb168c64893a3d91d28a/dclab/rtdc_dataset/check.py#L570). I also removed "user" from `OPTIONAL_KEYS` because I think it doesn't really fit there (and it being there doesn't do anything).
2. I have covered these now thanks. I see a `VisibleDeprecationWarning` from numpy: 
   ```
   dclab\dclab\definitions.py:407: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list- 
   or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 
   'dtype=object' when creating the ndarray
   data = np.array(data)
   ```
   But in our case, we are just creating one numpy array from an ndarray of length whatever, right? So we can ignore this?
   
   I had been using `pytest tests\test_file.py --disable-warnings`. I will now use `pytest tests\test_file.py -W ignore::DeprecationWarning`. It's those damn tensorflow warnings...

paulmueller reviewed (approved):
happy chap

paulmueller commented:
I simplified a few things and introduced the "config required" section in the docs and in the tests.

PinkShnack commented on docs/sec_av_feat_plugin.rst:
```diff
@@ -119,18 +119,17 @@ a feature is scalar or not (defaults to True).
 ------------------------------------------------
 In this :download:`example <data/example_plugin_metadata.py>`, the function
-:func:`compute_area_exponent` defines the basic feature `"area_exp"`,
+:func:`compute_area_exponent` defines the basic feature `area_exp`,
```
I would have thought here that the feature name is a string?

PinkShnack commented on docs/sec_av_feat_plugin.rst:
```diff
@@ -140,12 +139,21 @@ in the rtdc dataset's "user" configuration section.
-        In [4]: my_metadata = {"inlet": True, "n_channels": 4, "exp": 3}
+        # The plugin feature is not yet available, because "user:exp" is missing
+        In [4]: "area_exp" in ds
+        Out[4]: False
```
Cool,  I see now how much clearer this is.

PinkShnack commented on tests/test_plugin_features.py:
```diff
@@ -510,17 +513,20 @@ def test_pf_with_user_config_section():
-    with pytest.raises(KeyError):
+    with pytest.raises(KeyError,
+                       match=r"Feature \'area_of_region\' does not exist"):
         ds["area_of_region"]
```
Ah that's very useful (`match` I mean).
